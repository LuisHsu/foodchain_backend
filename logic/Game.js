import { shuffle } from '../utils/arr';
import { employees, houseTiles, mapTiles, marketingTiles, milestones } from '../gameConfigs';
import GameMap from './GameMap';
import { transformPosition } from '../utils/map';

export const setupMap = (playerNumber) => {
    const xSize = playerNumber >= 3 ? playerNumber : 3;
    const ySize = playerNumber >= 3 ? 4 : 3;
    const tileIds = shuffle(Object.keys(mapTiles)).slice(0, xSize * ySize);
    const configs = tileIds
        .map((tileId) => ({
            tileId,
            direction: Math.floor(Math.random() * 4),
        }));
    return new GameMap(xSize, ySize, configs);
};

export const setupPool = (playerNumber) => {
    const ads = { ...marketingTiles };
    if (playerNumber < 5) {
        delete ads.Mk16;
    }
    if (playerNumber < 4) {
        delete ads.Mk15;
    }
    if (playerNumber < 3) {
        delete ads.Mk12;
    }
    return { // An event may contain 1 or more keys defined below
        money: 50 * playerNumber,
        employees: Object.keys(employees)
            .map((id) => {
                let { amount } = employees[id];
                if (employees[id].limited) {
                    amount = playerNumber - 2;
                    if (amount === 0) {
                        amount = 1;
                    }
                }
                return ({
                    id,
                    amount,
                });
            })
            .reduce((map, e) => ({ ...map, [e.id]: e.amount }), {}),
        milestones: {
            remain: Object.keys(milestones),
            achieved: {},
            new: {},
        },
        houses: Object.keys(houseTiles).filter((key) => !houseTiles[key].fixed),
        gardenNumber: 8,
        ads: Object.keys(ads),
    };
};

export const generateWorkingOrder = (players) => {
    const order = [];
    for (let i = 0; i < players.length; i += 1) {
        order.push(i);
    }
    const randomOrder = shuffle(order);
    return randomOrder.map((i) => players[i]);
};

export default class Game {
    constructor(eventManager) {
        this.eventManager = eventManager;
    }

    initialize(players) {
        this.turn = 1;
        this.phase = 1;
        this.players = players;
        this.gameMap = setupMap(players.length);
        this.eventManager.notifyAll('setup_map', {
            tiles: this.gameMap.tileConfig,
        });

        this.pool = setupPool(players.length);
        this.eventManager.notifyAll('pool_update', this.pool);

        this.workingOrder = generateWorkingOrder(players);
        this.eventManager.notifyAll('order_decision', {
            available: [],
            selected: this.workingOrder,
        });

        this.restaurants = {};
        this.setupRestaurants();
    }

    setupRestaurants() {
        return new Promise((resolve) => {
            this.eventManager.on(
                'pick_first_restaurant',
                (username, msg) => this.onFirstRestaurantPick(username, [
                    msg.position.xTile,
                    msg.position.yTile,
                    msg.position.xSmall,
                    msg.position.ySmall,
                ], msg.direction, resolve),
            );
            const firstPick = this.workingOrder[this.workingOrder.length - 1];
            this.askFirstRestaurant(firstPick, true);
        });
    }

    askFirstRestaurant(username, skippable) {
        this.eventManager.notifyAll('first_restaurant_decision', {
            nextPlayer: username,
            skippable,
        });
    }

    onFirstRestaurantPick(username, position, direction, done) {
        if (!position) {
            this.restaurants[username] = null;
        } else {
            this.gameMap.putTile(username, position.map((p) => transformPosition(p, direction)));
            this.restaurants[username] = 1;
            this.eventManager.notifyAll('tiles_placement', {
                type: 'restaurant',
                position,
                direction,
            });
        }
        let nextIndex = (this.workingOrder.indexOf(username) - 1 + this.workingOrder.length);

        while (nextIndex >= 0 && this.restaurants[this.workingOrder[nextIndex % this.workingOrder.length]]) {
            nextIndex -= 1;
        }
        if (nextIndex < 0) {
            done?.();
        } else {
            const nextUser = this.workingOrder[nextIndex % this.workingOrder.length];
            this.askFirstRestaurant(nextUser, this.restaurants[nextUser] === undefined);
        }
    }
}
