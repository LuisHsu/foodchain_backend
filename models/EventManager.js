import jwt from 'jsonwebtoken';

export default class EventManager {
    constructor(io) {
        this.io = io;
        this.internalEventMap = {};
        this.socketEventMap = {};
        io.on('connection', (socket) => {
            try {
                const payload = jwt.verify(socket.handshake.auth.token, process.env.token_secret);
                console.log(`user: ${payload.username} -- connected`);
                this.internalEventMap.$$player_add?.(payload.username);
                Object.keys(this.socketEventMap)
                    .forEach((eventName) => {
                        socket.on(eventName, this.socketEventMap[eventName]);
                    });

                socket.on('disconnect', () => {
                    this.internalEventMap.$$player_remove?.(payload.username);
                    delete this.socketEventMap[socket];
                });
            } catch (e) {
                socket.disconnect();
            }
        });
    }

    notifyAll(eventName, payload) {
        this.io.emit(eventName, payload);
    }

    on(eventName, callback) {
        if (eventName.startsWith('$$')) {
            this.internalEventMap[eventName] = callback;
        } else {
            this.socketEventMap[eventName] = callback;
        }
    }
}
