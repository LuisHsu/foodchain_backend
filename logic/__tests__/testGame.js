import Game from '../Game';
import EventManager from '../../models/EventManager';

jest.mock('../../models/EventManager');

describe('test Game', () => {
    let eventMgr;
    beforeEach(() => {
        EventManager.mockClear();
        eventMgr = new EventManager();
    });
    describe('generateWorkingOrder', () => {
        it('success', () => {
            const game = new Game(eventMgr);
            game.generateWorkingOrder(['user1', 'user2', 'user3']);
            const mockEventMgr = EventManager.mock.instances[0];
            expect(mockEventMgr.notifyAll).toBeCalledWith('order_decision', {
                available: [],
                selected: expect.arrayContaining(['user1', 'user2', 'user3']),
            });
        });
    });

    describe('askFirstRestaurant', () => {
        it('not skippable', () => {
            const game = new Game(eventMgr);
            game.askFirstRestaurant('user1', false);
            const mockEventMgr = EventManager.mock.instances[0];
            expect(mockEventMgr.notifyAll).toBeCalledWith('first_restaurant_decision', {
                nextPlayer: 'user1',
                skippable: false,
            });
        });
        it('skippable', () => {
            const game = new Game(eventMgr);
            game.askFirstRestaurant('user1', true);
            const mockEventMgr = EventManager.mock.instances[0];
            expect(mockEventMgr.notifyAll).toBeCalledWith('first_restaurant_decision', {
                nextPlayer: 'user1',
                skippable: true,
            });
        });
    });
});
