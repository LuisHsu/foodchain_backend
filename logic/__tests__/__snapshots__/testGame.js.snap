// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Game generateWorkingOrder success 1`] = `
Array [
  "user2",
  "user3",
  "user1",
]
`;

exports[`test Game initialize normal 1`] = `
Array [
  Array [
    "setup_map",
    Object {
      "tiles": Array [
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 0,
          },
          "tileId": "M10",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 0,
          },
          "tileId": "M20",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 0,
          },
          "tileId": "M04",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 1,
          },
          "tileId": "M05",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 1,
          },
          "tileId": "M06",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 1,
          },
          "tileId": "M07",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 2,
          },
          "tileId": "M08",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 2,
          },
          "tileId": "M09",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 2,
          },
          "tileId": "M01",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 3,
          },
          "tileId": "M11",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 3,
          },
          "tileId": "M12",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 3,
          },
          "tileId": "M13",
        },
      ],
    },
  ],
  Array [
    "pool_update",
    Object {
      "ads": Array [
        "Mk01",
        "Mk02",
        "Mk03",
        "Mk04",
        "Mk05",
        "Mk06",
        "Mk07",
        "Mk08",
        "Mk09",
        "Mk10",
        "Mk11",
        "Mk12",
        "Mk13",
        "Mk14",
      ],
      "employees": Object {
        "E01": 12,
        "E02": 18,
        "E03": 12,
        "E04": 12,
        "E05": 12,
        "E06": 12,
        "E07": 12,
        "E08": 12,
        "E09": 6,
        "E10": 12,
        "E11": 1,
        "E12": 6,
        "E13": 6,
        "E14": 6,
        "E15": 6,
        "E16": 6,
        "E17": 6,
        "E18": 6,
        "E19": 6,
        "E20": 6,
        "E21": 6,
        "E22": 6,
        "E23": 1,
        "E24": 1,
        "E25": 1,
        "E26": 6,
        "E27": 3,
        "E28": 1,
        "E29": 1,
        "E30": 1,
        "E31": 1,
        "E32": 1,
      },
      "gardenNumber": 8,
      "houses": Array [
        "H01",
        "H03",
        "H06",
        "H09",
        "H11",
        "H14",
        "H17",
        "H19",
      ],
      "milestones": Object {
        "achieved": Object {},
        "new": Object {},
        "remain": Array [
          "Mi01",
          "Mi02",
          "Mi03",
          "Mi04",
          "Mi05",
          "Mi06",
          "Mi07",
          "Mi08",
          "Mi09",
          "Mi10",
          "Mi11",
          "Mi12",
          "Mi13",
          "Mi14",
          "Mi15",
          "Mi16",
          "Mi17",
          "Mi18",
        ],
      },
      "money": 150,
    },
  ],
  Array [
    "order_decision",
    Object {
      "available": Array [],
      "selected": Array [
        "user2",
        "user3",
        "user1",
      ],
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user1",
      "skippable": true,
    },
  ],
  Array [
    "tiles_placement",
    Object {
      "direction": 0,
      "position": Array [
        Array [
          0,
          0,
          3,
          3,
        ],
        Array [
          0,
          0,
          4,
          3,
        ],
        Array [
          0,
          0,
          3,
          4,
        ],
        Array [
          0,
          0,
          4,
          4,
        ],
      ],
      "type": "restaurant",
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user3",
      "skippable": true,
    },
  ],
  Array [
    "tiles_placement",
    Object {
      "direction": 0,
      "position": Array [
        Array [
          2,
          3,
          3,
          3,
        ],
        Array [
          2,
          3,
          4,
          3,
        ],
        Array [
          2,
          3,
          3,
          4,
        ],
        Array [
          2,
          3,
          4,
          4,
        ],
      ],
      "type": "restaurant",
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user2",
      "skippable": true,
    },
  ],
  Array [
    "tiles_placement",
    Object {
      "direction": 0,
      "position": Array [
        Array [
          0,
          3,
          3,
          3,
        ],
        Array [
          0,
          3,
          4,
          3,
        ],
        Array [
          0,
          3,
          3,
          4,
        ],
        Array [
          0,
          3,
          4,
          4,
        ],
      ],
      "type": "restaurant",
    },
  ],
]
`;

exports[`test Game initialize skip restaurant pick 1`] = `
Array [
  Array [
    "setup_map",
    Object {
      "tiles": Array [
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 0,
          },
          "tileId": "M10",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 0,
          },
          "tileId": "M20",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 0,
          },
          "tileId": "M04",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 1,
          },
          "tileId": "M05",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 1,
          },
          "tileId": "M06",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 1,
          },
          "tileId": "M07",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 2,
          },
          "tileId": "M08",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 2,
          },
          "tileId": "M09",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 2,
          },
          "tileId": "M01",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 3,
          },
          "tileId": "M11",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 3,
          },
          "tileId": "M12",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 3,
          },
          "tileId": "M13",
        },
      ],
    },
  ],
  Array [
    "pool_update",
    Object {
      "ads": Array [
        "Mk01",
        "Mk02",
        "Mk03",
        "Mk04",
        "Mk05",
        "Mk06",
        "Mk07",
        "Mk08",
        "Mk09",
        "Mk10",
        "Mk11",
        "Mk12",
        "Mk13",
        "Mk14",
      ],
      "employees": Object {
        "E01": 12,
        "E02": 18,
        "E03": 12,
        "E04": 12,
        "E05": 12,
        "E06": 12,
        "E07": 12,
        "E08": 12,
        "E09": 6,
        "E10": 12,
        "E11": 1,
        "E12": 6,
        "E13": 6,
        "E14": 6,
        "E15": 6,
        "E16": 6,
        "E17": 6,
        "E18": 6,
        "E19": 6,
        "E20": 6,
        "E21": 6,
        "E22": 6,
        "E23": 1,
        "E24": 1,
        "E25": 1,
        "E26": 6,
        "E27": 3,
        "E28": 1,
        "E29": 1,
        "E30": 1,
        "E31": 1,
        "E32": 1,
      },
      "gardenNumber": 8,
      "houses": Array [
        "H01",
        "H03",
        "H06",
        "H09",
        "H11",
        "H14",
        "H17",
        "H19",
      ],
      "milestones": Object {
        "achieved": Object {},
        "new": Object {},
        "remain": Array [
          "Mi01",
          "Mi02",
          "Mi03",
          "Mi04",
          "Mi05",
          "Mi06",
          "Mi07",
          "Mi08",
          "Mi09",
          "Mi10",
          "Mi11",
          "Mi12",
          "Mi13",
          "Mi14",
          "Mi15",
          "Mi16",
          "Mi17",
          "Mi18",
        ],
      },
      "money": 150,
    },
  ],
  Array [
    "order_decision",
    Object {
      "available": Array [],
      "selected": Array [
        "user2",
        "user3",
        "user1",
      ],
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user1",
      "skippable": true,
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user3",
      "skippable": true,
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user2",
      "skippable": true,
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user1",
      "skippable": false,
    },
  ],
]
`;

exports[`test Game setupPool 2 players 1`] = `
Object {
  "ads": Array [
    "Mk01",
    "Mk02",
    "Mk03",
    "Mk04",
    "Mk05",
    "Mk06",
    "Mk07",
    "Mk08",
    "Mk09",
    "Mk10",
    "Mk11",
    "Mk13",
    "Mk14",
  ],
  "employees": Object {
    "E01": 12,
    "E02": 18,
    "E03": 12,
    "E04": 12,
    "E05": 12,
    "E06": 12,
    "E07": 12,
    "E08": 12,
    "E09": 6,
    "E10": 12,
    "E11": 1,
    "E12": 6,
    "E13": 6,
    "E14": 6,
    "E15": 6,
    "E16": 6,
    "E17": 6,
    "E18": 6,
    "E19": 6,
    "E20": 6,
    "E21": 6,
    "E22": 6,
    "E23": 1,
    "E24": 1,
    "E25": 1,
    "E26": 6,
    "E27": 3,
    "E28": 1,
    "E29": 1,
    "E30": 1,
    "E31": 1,
    "E32": 1,
  },
  "gardenNumber": 8,
  "houses": Array [
    "H01",
    "H03",
    "H06",
    "H09",
    "H11",
    "H14",
    "H17",
    "H19",
  ],
  "milestones": Object {
    "achieved": Object {},
    "new": Object {},
    "remain": Array [
      "Mi01",
      "Mi02",
      "Mi03",
      "Mi04",
      "Mi05",
      "Mi06",
      "Mi07",
      "Mi08",
      "Mi09",
      "Mi10",
      "Mi11",
      "Mi12",
      "Mi13",
      "Mi14",
      "Mi15",
      "Mi16",
      "Mi17",
      "Mi18",
    ],
  },
  "money": 100,
}
`;

exports[`test Game setupPool 3 players 1`] = `
Object {
  "ads": Array [
    "Mk01",
    "Mk02",
    "Mk03",
    "Mk04",
    "Mk05",
    "Mk06",
    "Mk07",
    "Mk08",
    "Mk09",
    "Mk10",
    "Mk11",
    "Mk12",
    "Mk13",
    "Mk14",
  ],
  "employees": Object {
    "E01": 12,
    "E02": 18,
    "E03": 12,
    "E04": 12,
    "E05": 12,
    "E06": 12,
    "E07": 12,
    "E08": 12,
    "E09": 6,
    "E10": 12,
    "E11": 1,
    "E12": 6,
    "E13": 6,
    "E14": 6,
    "E15": 6,
    "E16": 6,
    "E17": 6,
    "E18": 6,
    "E19": 6,
    "E20": 6,
    "E21": 6,
    "E22": 6,
    "E23": 1,
    "E24": 1,
    "E25": 1,
    "E26": 6,
    "E27": 3,
    "E28": 1,
    "E29": 1,
    "E30": 1,
    "E31": 1,
    "E32": 1,
  },
  "gardenNumber": 8,
  "houses": Array [
    "H01",
    "H03",
    "H06",
    "H09",
    "H11",
    "H14",
    "H17",
    "H19",
  ],
  "milestones": Object {
    "achieved": Object {},
    "new": Object {},
    "remain": Array [
      "Mi01",
      "Mi02",
      "Mi03",
      "Mi04",
      "Mi05",
      "Mi06",
      "Mi07",
      "Mi08",
      "Mi09",
      "Mi10",
      "Mi11",
      "Mi12",
      "Mi13",
      "Mi14",
      "Mi15",
      "Mi16",
      "Mi17",
      "Mi18",
    ],
  },
  "money": 150,
}
`;

exports[`test Game setupPool 4 players 1`] = `
Object {
  "ads": Array [
    "Mk01",
    "Mk02",
    "Mk03",
    "Mk04",
    "Mk05",
    "Mk06",
    "Mk07",
    "Mk08",
    "Mk09",
    "Mk10",
    "Mk11",
    "Mk12",
    "Mk13",
    "Mk14",
    "Mk15",
  ],
  "employees": Object {
    "E01": 12,
    "E02": 18,
    "E03": 12,
    "E04": 12,
    "E05": 12,
    "E06": 12,
    "E07": 12,
    "E08": 12,
    "E09": 6,
    "E10": 12,
    "E11": 2,
    "E12": 6,
    "E13": 6,
    "E14": 6,
    "E15": 6,
    "E16": 6,
    "E17": 6,
    "E18": 6,
    "E19": 6,
    "E20": 6,
    "E21": 6,
    "E22": 6,
    "E23": 2,
    "E24": 2,
    "E25": 2,
    "E26": 6,
    "E27": 3,
    "E28": 2,
    "E29": 2,
    "E30": 2,
    "E31": 2,
    "E32": 2,
  },
  "gardenNumber": 8,
  "houses": Array [
    "H01",
    "H03",
    "H06",
    "H09",
    "H11",
    "H14",
    "H17",
    "H19",
  ],
  "milestones": Object {
    "achieved": Object {},
    "new": Object {},
    "remain": Array [
      "Mi01",
      "Mi02",
      "Mi03",
      "Mi04",
      "Mi05",
      "Mi06",
      "Mi07",
      "Mi08",
      "Mi09",
      "Mi10",
      "Mi11",
      "Mi12",
      "Mi13",
      "Mi14",
      "Mi15",
      "Mi16",
      "Mi17",
      "Mi18",
    ],
  },
  "money": 200,
}
`;

exports[`test Game setupPool 5 players 1`] = `
Object {
  "ads": Array [
    "Mk01",
    "Mk02",
    "Mk03",
    "Mk04",
    "Mk05",
    "Mk06",
    "Mk07",
    "Mk08",
    "Mk09",
    "Mk10",
    "Mk11",
    "Mk12",
    "Mk13",
    "Mk14",
    "Mk15",
    "Mk16",
  ],
  "employees": Object {
    "E01": 12,
    "E02": 18,
    "E03": 12,
    "E04": 12,
    "E05": 12,
    "E06": 12,
    "E07": 12,
    "E08": 12,
    "E09": 6,
    "E10": 12,
    "E11": 3,
    "E12": 6,
    "E13": 6,
    "E14": 6,
    "E15": 6,
    "E16": 6,
    "E17": 6,
    "E18": 6,
    "E19": 6,
    "E20": 6,
    "E21": 6,
    "E22": 6,
    "E23": 3,
    "E24": 3,
    "E25": 3,
    "E26": 6,
    "E27": 3,
    "E28": 3,
    "E29": 3,
    "E30": 3,
    "E31": 3,
    "E32": 3,
  },
  "gardenNumber": 8,
  "houses": Array [
    "H01",
    "H03",
    "H06",
    "H09",
    "H11",
    "H14",
    "H17",
    "H19",
  ],
  "milestones": Object {
    "achieved": Object {},
    "new": Object {},
    "remain": Array [
      "Mi01",
      "Mi02",
      "Mi03",
      "Mi04",
      "Mi05",
      "Mi06",
      "Mi07",
      "Mi08",
      "Mi09",
      "Mi10",
      "Mi11",
      "Mi12",
      "Mi13",
      "Mi14",
      "Mi15",
      "Mi16",
      "Mi17",
      "Mi18",
    ],
  },
  "money": 250,
}
`;
